# Describe dataset:
* ID: A unique identifier for the laptop.
* Company: The name of the company that produces the laptop.
* Product: The name of the laptop.
* TypeName: The type of laptop.
* Inches: The size of the screen in inches.
* ScreenResolution: The resolution of the screen.
* CPU: The laptop CPU.
* RAM: The amount of RAM in the laptop.
* Memory: The size of the hard drive.
* GPU: The graphics card name.
* OpSys: The name of the operating system.
* Weight: The laptop weight.
* Price: The price of the laptop.

### Reading the Inventory

Use the `CSV` module to read the laptop.csv file and separate the header from the rows

In [2]:
import csv

with open('laptop_price.csv',encoding='utf-8', errors='ignore') as f:
    reader = csv.reader(f)
    rows = list(reader)
    header = rows[0]
    rows = rows[1:]
    
print(header)
print(rows[:5])

['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']
[['1', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339.69'], ['2', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898.94'], ['3', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575.00'], ['4', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537.45'], ['5', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB SSD',

### Inventory Class


Start implementing a class to represent the inventory . It get the name of the CSV file as argument and reads it into **self.header** and **self.rows**.

In [9]:
class Inventory ():
    def __init__(self, csv_filename):
        with open(csv_filename,encoding='utf-8', errors='ignore') as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = float(row[-1])
            row[-1] = int(row[-1])
inventory = Inventory('laptop_price.csv')
print(inventory.header)
print(len(inventory.rows))

['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']
1303


### Finding a Laptop From the Id


Implement a get_laptop_from_id() function that given a laptop identifier find the row corresponding to that laptop.

In [27]:
class Inventory ():
    def __init__(self, csv_filename):
        with open(csv_filename,encoding='utf-8', errors='ignore') as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = float(row[-1])
            row[-1] = int(row[-1])
    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]== laptop_id:
                return row
        return None

In [28]:
inventory = Inventory('laptop_price.csv')
print(inventory.get_laptop_from_id('37'))
print(inventory.get_laptop_from_id('36'))

['37', 'Acer', 'Aspire 3', 'Notebook', '15.6', '1366x768', 'Intel Core i3 7130U 2.7GHz', '4GB', '1TB HDD', 'Intel HD Graphics 620', 'Linux', '2.1kg', 367]
['36', 'Lenovo', 'IdeaPad 120S-14IAP', 'Notebook', '14', '1366x768', 'Intel Celeron Dual Core N3350 1.1GHz', '4GB', '64GB Flash Storage', 'Intel HD Graphics 500', 'Windows 10', '1.44kg', 249]


### Improving ID Lookups]


Improve the time complexity of finding a laptop with a given ib by precomputing a dictionary that maps laptop ids to rows.

In [52]:
class Inventory ():
    def __init__(self, csv_filename):
        with open(csv_filename,encoding='utf-8', errors='ignore') as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = float(row[-1])
            row[-1] = int(row[-1])
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row

In [43]:
inventory = Inventory('laptop_price.csv')
print(inventory.get_laptop_from_id_fast('37'))
print(inventory.get_laptop_from_id_fast('36'))

['37', 'Acer', 'Aspire 3', 'Notebook', '15.6', '1366x768', 'Intel Core i3 7130U 2.7GHz', '4GB', '1TB HDD', 'Intel HD Graphics 620', 'Linux', '2.1kg', 367]
['36', 'Lenovo', 'IdeaPad 120S-14IAP', 'Notebook', '14', '1366x768', 'Intel Celeron Dual Core N3350 1.1GHz', '4GB', '64GB Flash Storage', 'Intel HD Graphics 500', 'Windows 10', '1.44kg', 249]



### Comparing the Performance

In [59]:
import time
import random

ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]

inventory = Inventory('laptop_price.csv')

total_time_no_dict = 0
for identifier in ids:
    start = time.time()
    inventory.get_laptop_from_id(identifier)
    end = time.time()
    total_time_no_dict += end - start 
    
total_time_dict = 0
for identifier in ids:
    start = time.time()
    inventory.get_laptop_from_id_fast(identifier)
    end = time.time()
    total_time_dict += end - start
    
    
print('No Dictionary Total Time:' + str(total_time_no_dict))
print('Dictionary Time: ' + str(total_time_dict))

No Dictionary Total Time:0.37648749351501465
Dictionary Time: 0.012545347213745117


### Analysis


The get_laptop_from_id() method has time complexity O(R) where R is the number of rows. In contrast, get_laptop_from_id_fast() has time complexity O(1).

We got: 
No Dictionary Total Time:0.37648749351501465
Dictionary Time: 0.012545347213745117

We can see a significant improve in performance. If we divide 0.376 by 0.012 we see that the new method is about 31 times faster for this input size.


### Two Laptop Promotion

Write a method that finds whether we can spend a given amount of money by purchasing either one or two laptops.

In [64]:
class Inventory():
    def __init__(self, csv_filename):
        with open(csv_filename,encoding='utf-8', errors='ignore') as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = float(row[-1])
            row[-1] = int(row[-1])
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False
            

In [65]:
inventory = Inventory('laptop_price.csv')
print(inventory.check_promotion_dollars(1000))
print(inventory.check_promotion_dollars(442))

True
False


### Optimizing Laptop Premium

In [92]:
class Inventory():
    def __init__(self, csv_filename):
        with open(csv_filename,encoding='utf-8', errors='ignore') as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = float(row[-1])
            row[-1] = int(row[-1])
        
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row

        self.prices = set()
        for row in rows:
            self.prices.add(row[-1])

    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False
    
    def check_promotion_dollars_fast(self,dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if dollars - int(price) in self.prices:
                return True
        return False

In [93]:
inventory = Inventory('laptop_price.csv')

print(inventory.check_promotion_dollars_fast(1000))

True


### Compaing Promotion Functions

Compare the performance of both methods for the promotion.

In [ ]:
prices = [random.randint(100, 5000) for _ in range(100)] 

inventory = Inventory('laptop_price.csv')

total_time_no_set = 0
for value in prices:
    start = time.time()
    inventory.check_promotion_dollars(value)
    end = time.time()
    total_time_no_set += end - start
    
total_time_set = 0
for value in prices:
    start = time.time()
    inventory.check_promotion_dollars_fast(value)
    end = time.time()
    total_time_set += end - start
    
print(total_time_no_set)

print(total_time_set)

### Finding Laptops Within a Budget

In [ ]:
def row_price(row):
    return row[-1]

class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
            
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
            
        self.rows_by_price = sorted(self.rows, key=row_price)
            
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        
        for row1 in self.rows:                    
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False   
    
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if dollars - price in self.prices:
                return True
        return False
    
    def find_laptop_with_price(self, target_price):
        range_start = 0                                   
        range_end = len(self.rows_by_price) - 1                       
        while range_start < range_end:
            range_middle = (range_end + range_start) // 2  
            value = self.rows_by_price[range_middle][-1]
            if value == target_price:                            
                return range_middle                        
            elif value < target_price:                           
                range_start = range_middle + 1             
            else:                                          
                range_end = range_middle - 1 
        if self.rows_by_price[range_start][-1] != target_price:                  
            return -1                                      
        return range_start